### DISCUSSION TITLE
Comparison of released Coqui-TTS models

### DISCUSSION THREAD

**Discussion by mbarnig** [coqui-ai/TTS#522] [Show and tell]
I tested the working and performance of the english, french and german Coqui-TTS models in a same environment : the docker container [synesthesiam/coqui-tts-0.0.13.2](https://hub.docker.com/layers/synesthesiam/coqui-tts/0.0.13.2/images/sha256-ce0910e0bd327cc8893af54d58b269b6048f1ad5b284c8b0dcd84636aef932a0?context=explore) created two weeks ago. I used the linguistic test sentence [The Northwind and the Sun](https://www.phonetics.expert/north-wind-and-the-sun) in the related languages. Here are my findings: 

## CPU
model       | processing | proc. time | rt. factor | problems
------------|------------|------------|------------|---------
en-ek1 | ko | 164,29 s | 24,44  | decoder stopped with max decoder-steps
en-taco-dcc | OK | 2,66 s |0,37  | none
en-sam  | OK |  2,81 s |  0,35  | none 
en-taco-dca | OK | 1,37 s | 0,20  | none
en-glow | OK | 0,59 s | 0,07  | none
en-speedy | OK |  0,61 s | 0,08  | none
 fr-mai | ko | 3,64 s | 0,49 | sound stops in the middle of the sentence
de-thorsten | OK | 245,26 s | 22,40  | very high processing time
en-vctk | ko | - | - | JSONDecoderError

You can listen to the sound samples and view the related waveforms and test-sentences on my website [Internet with a Brain](https://www.web3.lu/text-to-speech-sound-samples-from-coqui-tts/).

In a second step I did the same tests with the [GPU Docker container]( https://hub.docker.com/layers/synesthesiam/coqui-tts/gpu-0.0.13.2/images/sha256-beca2f15d70cb29207d55f21ba75bd4f7250f483fe944b4a430a4200991ac98f?context=explore). I expected to get lower processing times, but they were more than twice as high. I created my own Docker file and also run the tests in my native Ubuntu20.04 environment. The resulting processing times were always of the same order.

To exclude a wrong configuration of my NVIDIA GeForce RTX2070 Super graphic card in my PC, I repeated my tests in Google Colab. I set the runtime to GPU and programmed some cells to verify the correct cuda usage and to get the name of the GPU : > Tesla 4. 

I installed first the Coqui-TTS release 0.0.13.2 from the Python Package Index. In a second trial I cloned the GitHub Coqui-TTS repository, followed by a checkout for version 0.0.13 and by an installation with pip. espeak-ng was installed separately. Both Coqui-TTS installations were succesful.  

I created the following cell to create a proxy in Colab for the Coqui-TTS server :
```
from google.colab.output import eval_js
print(eval_js("google.colab.kernel.proxyPort(5002)"))
```   
I used the following command to run the TTS server in the folder `/usr/local/lib/python3.7/dist-packages/TTS/server/` :   
```   
!CUDA_VISIBLE_DEVICES="0" python server.py --model_name <model-name from --list_models> --use_cuda True  --debug True
``` 
Example :   
```
!CUDA_VISIBLE_DEVICES="0" python server.py --model_name tts_models/en/ljspeech/tacotron2-DDC --use_cuda True  --debug True
``` 

For the installed version I used the command
```   
!CUDA_VISIBLE_DEVICES="0" tts-server <model-name from --list_models> --use_cuda True  --debug True
```
Example :
```    
!CUDA_VISIBLE_DEVICES="0" tts-server --model_name tts_models/en/vctk/sc-glow-tts --use_cuda True  --debug True
```  
I concluded that eventually the interworking of the server.py code with the GPU is not correct and that the higher processing times are caused by the cuda overhead . Therefore I tried also the following command in the folder `/usr/local/lib/python3.7/dist-packages/TTS/bin/` :   
```
!python synthesize.py --text "<utterance>" --model_name <model-name from --list_models>  --out_path test.wav
```    
Example :
```   
!python synthesize.py --text "The North Wind and the Sun were disputing which was the stronger, when a traveler came along wrapped in a warm cloak." --model_name tts_models/en/ljspeech/tacotron2-DDC --use_cuda True --out_path en-taco2-dcc.wav
```
respectively :   

```   
tts --text "Text for TTS" \
    --model_name "<type>/<language>/<dataset>/<model_name>" \
    --out_path test.wav
```  
 With these commands the processing times are lower, except for the glow-tts and speedy models.
 
 Here are the results :
 ## GPU
model       | processing | proc. time | rt. factor | problems
------------|------------|------------|------------|---------
en-ek1 | OK | 8,51 s | 1,27  | high processing time
en-taco-dcc | OK | 0,71 s |0,10  | none
en-sam  | OK |  1,11 s |  0,14  | none 
en-taco-dca | OK | 0,76 s | 0,11  | none
en-glow | OK | 1,36 s | 0,17  | none
en-speedy | OK |  0,81 s | 0,11  | none
fr-mai | ko | 1,15 s | 0,16 | sound stops in the middle of the sentence
de-thorsten | OK | 13,53 s | 1,24  | high processing time
en-vctk | ko | - | - | JSONDecoderError

The models `en-ek1` and `de-thorsten` have high processing times in all cases. The sound of the `fr-mai` model stops always in the middle of the sentence.

To play sounds in Colab I use the following cell-script :  
```
import librosa
import IPython.display as ipd
x, sr = librosa.load('tts-de-output.wav')
ipd.Audio(x, rate=sr)
```    

The model `en-vctk` generates the same JSONDecoderError in all cases, even with the simple command 
`!python synthesize.py --model_name tts_models/en/vctk/sc-glow-tts --use_cuda True --list_speaker_idxs` 

Here are the logs from colab :
```   
> tts_models/en/vctk/sc-glow-tts is already downloaded.
Traceback (most recent call last):
  File "synthesize.py", line 265, in <module>
    main()
  File "synthesize.py", line 234, in main
    args.use_cuda,
  File "/usr/local/lib/python3.7/dist-packages/TTS/utils/synthesizer.py", line 72, in __init__
    self._load_tts(tts_checkpoint, tts_config_path, use_cuda)
  File "/usr/local/lib/python3.7/dist-packages/TTS/utils/synthesizer.py", line 117, in _load_tts
    self.tts_config = load_config(tts_config_path)
  File "/usr/local/lib/python3.7/dist-packages/TTS/utils/io.py", line 50, in load_config
    data = read_json_with_comments(config_path)
  File "/usr/local/lib/python3.7/dist-packages/TTS/utils/io.py", line 33, in read_json_with_comments
    data = json.loads(input_str)
  File "/usr/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.7/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Invalid control character at: line 43 column 21 (char 2063)
``` 

Your questions and comments are welcome.

**Comment by thorstenMueller** [comment]
Hello @mbarnig  and thanks for your great comparison :blush:.

My "de-thorsten" model uses WaveGrad vocoder which is known for a high real time factor (_= high  processing time_).
I'm training a Fullband-MelGAN vocoder for my model right now (_training is at step 814k_). I plan to release it after reaching 1mio. steps. This model will have a nice RTF from around 0,5. This should reduce required processing time significantly.

As you've written here: https://www.web3.lu/text-to-speech-sound-samples-from-coqui-tts/
>> I will fulfill my dream to develop a high quality luxembourgish speech synthesizer, with the tools and kind help of these communities. 

Today i'm speaking on my experiences while recording my open german voice dataset (https://github.com/thorstenMueller/deep-learning-german-tts/) on Youtube.
Maybe this could be interesting for you:

https://www.youtube.com/watch?v=m-Uwb-Bg144

**Reply by mbarnig** [reply]
@thorstenMueller : thank you for the reminder. I will view with great pleasure your presentation on Youtube this evening.

I look forward to test your model with the new vocoder.

**Reply by mbarnig** [reply]
@thorstenMueller : Danke für die interessante Presentation auf Youtube. Sehr guter Inhalt, ausgezeichneter, lebendiger und humorvoller Vortrag.
![youtube-thumnail-2](https://user-images.githubusercontent.com/1360633/120538457-d2b67780-c3e6-11eb-9ae4-c6a0566366bb.png)

**Comment by erogol** [comment]
This is awesome. Thanks for this comparison @mbarnig.

Also thanks for mentioning Coqui in your post.

One question, how did you compute the real-time factor?

**Reply by mbarnig** [reply]
I took the values for the `Processing Time` and the `Real-time factor` from the logs of the model inference run.
  
Here is the example for running the french model in Colab :
```   
!python synthesize.py --text "La bise et le soleil se disputaient, chacun assurant qu'il était le plus fort, quand ils ont vu un voyageur qui s'avançait, enveloppé dans son manteau." --model_name tts_models/fr/mai/tacotron2-DDC --use_cuda True --out_path tts-fr-output.wav
 > Downloading model to /root/.local/share/tts/tts_models--fr--mai--tacotron2-DDC
 > Downloading model to /root/.local/share/tts/vocoder_models--universal--libri-tts--fullband-melgan
 > Using model: Tacotron2
 > Generator Model: fullband_melgan_generator
 > Text: La bise et le soleil se disputaient, chacun assurant qu'il était le plus fort, quand ils ont vu un voyageur qui s'avançait, enveloppé dans son manteau.
 > Text splitted to sentences.
["La bise et le soleil se disputaient, chacun assurant qu'il était le plus fort, quand ils ont vu un voyageur qui s'avançait, enveloppé dans son manteau."]
 > interpolating tts model output.
 > before interpolation : (80, 567)
 > after interpolation : torch.Size([1, 80, 850])
 > Processing time: 1.1658344268798828
 > Real-time factor: 0.15701473762692025
 > Saving output to tts-fr-output.wav
 ```   
 
 Here is the example for running the english glow model in server-mode in the Docker CPU container :  
```    
root@c2f8c20b3a95:/workspace# tts-server --model_name tts_models/en/ljspeech/glow-tts  --use_cuda true --debug true
 > Downloading model to /root/.local/share/tts/tts_models--en--ljspeech--glow-tts
 > Downloading model to /root/.local/share/tts/vocoder_models--en--ljspeech--multiband-melgan
 > Using model: glow_tts
 > Generator Model: multiband_melgan_generator
 * Serving Flask app "TTS.server.server" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
[INFO]  * Running on http://[::]:5002/ (Press CTRL+C to quit)
[INFO]  * Restarting with stat
 > tts_models/en/ljspeech/glow-tts is already downloaded.
 > vocoder_models/en/ljspeech/multiband-melgan is already downloaded.
 > Using model: glow_tts
 > Generator Model: multiband_melgan_generator
[WARNING]  * Debugger is active!
[INFO]  * Debugger PIN: 315-458-863
[INFO] ::ffff:172.17.0.1 - - [28/May/2021 14:03:12] "GET / HTTP/1.1" 200 -
[INFO] ::ffff:172.17.0.1 - - [28/May/2021 14:03:12] "GET /static/coqui-log-green-TTS.png HTTP/1.1" 304 -
 > Model input: The North Wind and the Sun were disputing which was the stronger, when a traveler came along wrapped in a warm cloak.
 > Text splitted to sentences.
['The North Wind and the Sun were disputing which was the stronger, when a traveler came along wrapped in a warm cloak.']
 > Processing time: 0.7767393589019775
 > Real-time factor: 0.09488909928080735
[INFO] ::ffff:172.17.0.1 - - [28/May/2021 14:03:15] "GET /api/tts?text=The%20North%20Wind%20and%20the%20Sun%20were%20disputing%20which%20was%20the%20stronger%2C%20when%20a%20traveler%20came%20along%20wrapped%20in%20a%20warm%20cloak.&speaker_id=&style_wav= HTTP/1.1" 200 -
``` 

The values for the `Processing Time` and the `Real-time factor` change at each run in a range of a few %.

I will repeat the Colab-GPU tests with the `synthesize.py` script in my Docker CPU and GPU containers to complete my overview.

@erogol : Should I open an issue for the VCTK model (JSONDecoderError) and for the french model (sound cut in the middle of the sentence) ?

**Reply by mbarnig** [reply]
@erogol : Thank you for editing my results table to bring it in the correct format.

**Comment by mbarnig** [comment]
The missing GPU support for the Coqui-TTS server was fixed with [commit b8b79a5](https://github.com/coqui-ai/TTS/commit/b8b79a5e5a1f175680a63539cd4235f18cb3ead8). I applied this change in version 0.0.13.2 and repeated my comparison of the released english, french and german models in a Colab notebook, now with GPU Runtime. The broken multispeaker model vctk was also working as expected.

Here are the values for the processing times and real-time factors printed in the logs. I used this time the full text (6 sentences) of the **North Wind & Sun** fable. Therefore we can't compare these values with those of my first post.

## Coqui-TTS server running in Colab with GPU Runtime
model       | processing | proc. time | rt. factor | problems
------------|------------|------------|------------|---------
en-ek1 | OK| 40,04 s | 1,17 | high processing time
en-taco-ddc | OK | 0,18 s |0,09  | none
en-sam  | OK |  5,14 s |  0,13  | none 
en-taco-dca | OK | 3,72 s | 0,10  | none
en-glow | OK | 3,07 s | 0,08  | none
en-speedy | OK |  3,60 s | 0,09  | none
fr-mai | ko | 5,67 s | 0,16 | speech stops
de-thorsten | OK | 52,85 s | 1,23  | high processing time
de-thorsten-multiband-melgan-vocoder | OK | 6,97 s | 0,16 | none
nl-taco-ddc | OK | 4,93 s | 0,14 | none
cn-taco-ddc | OK | 2,52 s | 0,07 | none
en-vctk-p260 | OK | 15,70 s | 0,52  | none
en-vctk-p267 | OK | 16,79 s | 0,54  | none
en-vctk-p301 | OK | 18,47 s | 0,51  | none

You can read the fables and listen to the soundfiles in this [GitHub webpage](https://mbarnig.github.io/TTS-Models-Comparison/).

@thorstenMueller : I tested your model with the multiband-melgan-vocoder.

**Reply by erogol** [reply]
awesome !! Thx for the update.
